DRV2605 DT Property Cleanup

rriveramcrus opened the pull request:
https://github.com/zephyrproject-rtos/zephyr/pull/76343#discussion_r1713519188
Address the issues raised by this comment:
- Drop manufacturer prefix from properties in the driver
- Document the GPIOs in the devicetree bindings

rriveramcrus committed:
dts: haptics: drv2605: Documents GPIOs

Documents the enable GPI and the input/trigger GPI in the dt bindings.

Signed-off-by: Ricardo Rivera-Matos <rriveram@opensource.cirrus.com>

rriveramcrus committed:
drivers: haptics: drv2605: Drops ti prefix from DT props

Removes the "ti_" prefix form the the device tree properties.

Signed-off-by: Ricardo Rivera-Matos <rriveram@opensource.cirrus.com>

fabiobaltieri reviewed (commented):
`drv2605_pm_state` also seems unused, also should this allow setting DRV2605_REG_RATED_VOLTAGE and DRV2605_REG_OVERDRIVE_CLAMP_VOLTAGE? I think they are fairly important parameters

rriveramcrus committed:
drivers: haptics: drv2605: Drops unused drv2605_pm_state

Removes an unused enum defining the DRV2605 power states. This enum is
dead code so it is being removed.

Signed-off-by: Ricardo Rivera-Matos <rriveram@opensource.cirrus.com>

kartben commented on drivers/haptics/drv2605.c:
```diff
@@ -601,11 +601,11 @@ static const struct haptics_driver_api drv2605_driver_api = {
-		.feedback_brake_factor = DT_INST_ENUM_IDX_OR(inst, ti_feedback_brake_factor, 3),   \
-		.loop_gain = DT_INST_ENUM_IDX_OR(inst, ti_loop_gain, 2),                           \
-		.actuator_mode = DT_INST_ENUM_IDX_OR(inst, ti_actuator_mode, 0),                   \
+		.en_gpio = GPIO_DT_SPEC_INST_GET_OR(inst, en_gpios, {}),                           \
+		.in_trig_gpio = GPIO_DT_SPEC_INST_GET_OR(inst, in_trig_gpios, {}),                 \
+		.feedback_brake_factor = DT_INST_ENUM_IDX_OR(inst, feedback_brake_factor, 3),      \
+		.loop_gain = DT_INST_ENUM_IDX_OR(inst, loop_gain, 2),                              \
+		.actuator_mode = DT_INST_ENUM_IDX_OR(inst, actuator_mode, 0),                      \
```
The default values should actually be set (and maybe more importantly, documented) in the binding... I think?

kartben replied:
(I know this is not strictly related to this change, but I missed it in initial review and am catching this now as I'm actually starting to play with the driver). Thanks!

rriveramcrus replied:
> The default values should actually be set (and maybe more importantly, documented) in the binding... I think?

Will do!

> as I'm actually starting to play with the driver

Don't have too much fun buzzing around.

rriveramcrus committed:
drivers: haptics: drv2605: Introduces rated voltage prop

Adds support for boot time configuration of the rated voltage at
initialization via "vib-rated-mv" devicetree property.

Signed-off-by: Ricardo Rivera-Matos <rriveram@opensource.cirrus.com>

rriveramcrus committed:
drivers: haptics: drv2605: Introduces overdrive clamp prop

Adds "vib-overdrive-mv" device tree property to configure the overdrive
clamp at initialization.

Signed-off-by: Ricardo Rivera-Matos <rriveram@opensource.cirrus.com>

rriveramcrus committed:
drivers: haptics: drv2605: Pick up property defaults from DT

Adds the default value for the existing properties to the device tree
and fall back on those values at initialization if necessary.

Signed-off-by: Ricardo Rivera-Matos <rriveram@opensource.cirrus.com>

rriveramcrus commented:
@fabiobaltieri 
> `drv2605_pm_state` also seems unused, 

Will drop, good catch.

> also should this allow setting DRV2605_REG_RATED_VOLTAGE and DRV2605_REG_OVERDRIVE_CLAMP_VOLTAGE? I think they are fairly important parameters

Sure, ack!

kartben commented on dts/bindings/haptics/ti,drv2605.yaml:
```diff
@@ -26,10 +27,29 @@ properties:
       - "8X"
       - "16X"
       - "DISABLED"
+    default: "4X"
```
Please document default values 
https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#rules-for-default-values

kartben commented on dts/bindings/haptics/ti,drv2605.yaml:
```diff
@@ -26,10 +27,29 @@ properties:
+    default: "MEDIUM"
+  rated-voltage:
+    type: int
+    default: 0x3f
+    description: |
+      Sets the reference voltage for full-scale output during closed-loop
+      operation.
+  overdrive-clamp:
+    type: int
+    default: 0x89
```
I'm not sure it's desirable to have the raw values there, i.e it is very opaque and hard to understand how this is a "voltage"
Could this be expressed in mV and the actual values for OD_CLAMP etc. automatically computed? It seems to be how Linux does it. Also would be good to use same property names I guess (`vib-overdrive-mv`, `vib-rated-mv`)?

fabiobaltieri replied:
yeah ideally the driver would do the calculations, they are a bit tricky though as the meaning changes depending on the mode

did not realize that there was a Linux driver though, may be worth checking it out

rriveramcrus replied:
``` c
/*
 * Rated and Overdriver Voltages:
 * Calculated using the formula r = v * 255 / 5.6
 * where r is what will be written to the register
 * and v is the rated or overdriver voltage of the actuator
 */
static int drv260x_calculate_voltage(unsigned int voltage)
{
	return (voltage * 255 / 5600);
}
```
From the Linux driver, I see this is how they're converting between raw codes and voltages. It's not clear to me from the datasheet where this comes from.
![image](https://github.com/user-attachments/assets/23b52a1f-a83d-4aab-b3bd-486c7a0a75bd)

The datasheet walks the user to a raw code based on the actuator's electrical characteristics rather than provide a direct conversion.

![image](https://github.com/user-attachments/assets/f1afcdb3-abd0-4c6c-b6b7-74e1742faa2f)
I do see this though. Playing pretend chip designer I guess all of these voltage thresholds are tied internally to the same reference, so it would make sense for `RATED_VOLTAGE` and `OD_CLAMP` to scale the same way.

Having worked with the author of the Linux driver, Dan, at TI I do feel confident that the equation from the c snippet is correct. I'll go ahead and convert these properties to natural units.

fabiobaltieri replied:
thanks for doing the investigation work

kartben reviewed (changes requested):
Hopefully @fabiobaltieri can chime in an confirm I am not making unreasonable requests 😅

BTW the MikroE Vibro 4 click I'm using seems to use a motor that requires LRA open-loop mode. No idea what this actually means 😅 but it's true the motor vibrates more smoothly when I hard code `CTRL3.LRA_OPEN_LOOP=1`. It feels like there are many other flags from CTRL1/2/3 that would be worth exposing, and I wonder what's the best way to do this, since there's certainly *a lot* of things that can be tuned there (so to be clear: I'm not asking you to add them in this PR). Regarding open loop specifically, it feels like having an `open-loop` boolean that sets either `LRA_OPEN_LOOP` or `ERM_OPEN_LOOP` based on the current mode would be a good starting point?

Thanks!

rriveramcrus committed:
dts: bindings: drv2605: Requires actuator-mode prop

Requires the actuator-mode property to be set as there is no safe
default value.

Signed-off-by: Ricardo Rivera-Matos <rriveram@opensource.cirrus.com>

rriveramcrus committed:
drivers: haptics: drv2605: Hard codes open loop mode for LRA mode

Enables open loop operation when initializing into LRA mode. Open loop
operation is enabled by default for ERM mode. This is hard coded for LRA
mode because calibration is currently unsupported.

Signed-off-by: Ricardo Rivera-Matos <rriveram@opensource.cirrus.com>

rriveramcrus commented:
@kartben 
> I'm using seems to use a motor that requires LRA open-loop mode. No idea what this actually means 😅 but it's true the motor vibrates more smoothly when I hard code CTRL3.LRA_OPEN_LOOP=1

I will hardcode this in the driver based on the actuator mode selected. I am opposed to adding a property to control open/closed loop operation without adding support for calibration and OTP burning. 

My DRV2605 breakout board does not make it easy to swap out actuators (it's cheap) and adding calibration support would require testing a couple different actuators. I defer adding calibration and OTP burning to a future PR.

fabiobaltieri commented on drivers/haptics/drv2605.c:
```diff
@@ -600,6 +609,7 @@ static const struct haptics_driver_api drv2605_driver_api = {
 		.feedback_brake_factor = DT_INST_ENUM_IDX(inst, feedback_brake_factor),            \
 		.loop_gain = DT_INST_ENUM_IDX(inst, loop_gain),                                    \
 		.actuator_mode = DT_INST_ENUM_IDX(inst, actuator_mode),                            \
+		.rated_voltage = DRV2605_CALCULATE_VOLTAGE(DT_INST_PROP(inst, vib_rated_mv)),      \
```
would use `_mv` in the field name as well (`rated_voltage_mv`), same in the next one

rriveramcrus replied:
After `DRV2605_CALCULATE_VOLTAGE` expands the value is no longer in millivolts. I feel adding the unit suffix is misleading.

fabiobaltieri replied:
oh I see, sorry I missed that bit

fabiobaltieri commented on drivers/haptics/drv2605.c:
```diff
@@ -124,11 +127,9 @@ LOG_MODULE_REGISTER(DRV2605, CONFIG_HAPTICS_LOG_LEVEL);
-};
+#define DRV2605_VOLTAGE_SCALE_FACTOR_MV 5600
+
+#define DRV2605_CALCULATE_VOLTAGE(_volt) (_volt * 255) / DRV2605_VOLTAGE_SCALE_FACTOR_MV
```
```suggestion
#define DRV2605_CALCULATE_VOLTAGE(_volt) ((_volt * 255) / DRV2605_VOLTAGE_SCALE_FACTOR_MV)
```

rriveramcrus replied:
ack

fabiobaltieri commented:
Hey played a bit with the driver in this PR, I have an unknown LRA salvaged from a phone connected, seems to work quite nicely, just the comment above about the macro.

kartben commented on drivers/haptics/drv2605.c:
```diff
@@ -458,6 +459,25 @@ static int drv2605_hw_config(const struct device *dev)
+		return ret;
+	}
+
+	if (config->actuator_mode == DRV2605_ACTUATOR_MODE_LRA) {
+		ret = i2c_reg_update_byte_dt(&config->i2c, DRV2605_REG_CONTROL3,
+					     DRV2605_LRA_OPEN_LOOP, DRV2605_LRA_OPEN_LOOP);
+		if (ret < 0) {
+			return ret;
+		}
+	}
```
should this set `DRV2605_ERM_OPEN_LOOP` when in `MODE_ERM`?
can be a follow-up PR if this makes sense, and please ignore if it doesn't :)

rriveramcrus replied:
`ERM_OPEN_LOOP` is set by default at power-on for the DRV2605 so I think it is safe to ignore for now. Once calibration support is added, the driver will need to make a decision whether to override the default if appropriate.
